clang/lib/AST/Decl.cpp

```diff
@@ -3292,11 +3292,9 @@ bool FunctionDecl::isImmediateFunction() const {
 }
 
 bool FunctionDecl::isMain() const {
-  const TranslationUnitDecl *tunit =
-    dyn_cast<TranslationUnitDecl>(getDeclContext()->getRedeclContext());
-  return tunit &&
-         !tunit->getASTContext().getLangOpts().Freestanding &&
-         isNamed(this, "main");
+  const DeclContext *DC = getDeclContext();
+  return isNamed(this, "main") && !getLangOpts().Freestanding &&
+         (DC->getRedeclContext()->isTranslationUnit() || DC->isLinkageSpec());
```

Comment:
```suggestion
         (DC->getRedeclContext()->isTranslationUnit() || isExternC());
```

I think this should do the trick. Also please add a test for
```c++
namespace foo { 
    extern "C++" void main() {} 
}
```
which is horrible, but valid.